### PR TITLE
support ssl-hostname-overwrite

### PR DESCRIPTION
**Pull request by davidkhala**
Signed-off-by: DavidLiu <david.yx.liu@oracle.com>

### CODE REVIEW

**Comment by davidkhala** [comment]
resolve a point on https://github.com/Hyperledger-TWGC/tape/issues/222

File: pkg/infra/basic/client.go [hunk 1]
```diff
@@ -81,7 +81,10 @@ func DialConnection(node Node, logger *log.Logger) (*grpc.ClientConn, error) {
 	var connError error
 	var conn *grpc.ClientConn
 	for i := 1; i <= 3; i++ {
-		conn, connError = gRPCClient.NewConnection(node.Addr, func(tlsConfig *tls.Config) { tlsConfig.InsecureSkipVerify = true })
+		conn, connError = gRPCClient.NewConnection(node.Addr, func(tlsConfig *tls.Config) {
+			tlsConfig.InsecureSkipVerify = true
+			tlsConfig.ServerName = node.SslTargetNameOverride
```
Comment (SamYuan1990) [inline pkg/infra/basic/client.go:86]: what if SslTargetNameOverride leave to empty?

File: pkg/infra/basic/config.go [hunk 2]
```diff
@@ -61,14 +61,15 @@ type Config struct {
...
-	Addr          string `yaml:"addr"`
-	TLSCACert     string `yaml:"tls_ca_cert"`
-	Org           string `yaml:"org"`
-	TLSCAKey      string `yaml:"tls_ca_key"`
-	TLSCARoot     string `yaml:"tls_ca_root"`
-	TLSCACertByte []byte
-	TLSCAKeyByte  []byte
-	TLSCARootByte []byte
+	Addr                  string `yaml:"addr"`
+	SslTargetNameOverride string `yaml:"ssl_target_name_override"`
```
Comment (SamYuan1990) [inline pkg/infra/basic/config.go:65]: need a sample in readme document for usage

File: pkg/infra/basic/client.go (hunk 1 above)
Comment (SamYuan1990) [reply pkg/infra/basic/client.go:86]: and I am not sure if we should add any kind of test as coverage?

Comment (davidkhala) [reply pkg/infra/basic/client.go:86]: option.servername is a string so should be default to empty string

Good idea to have test coverage.
I will migrate my test back.

File: pkg/infra/basic/config.go (hunk 2 above)
Comment (davidkhala) [reply pkg/infra/basic/config.go:65]: Where is the original document for struct Node?

Comment (SamYuan1990) [reply pkg/infra/basic/config.go:65]: https://github.com/Hyperledger-TWGC/tape/blob/master/docs/configfile.md

Comment (davidkhala) [reply pkg/infra/basic/config.go:65]: 有ssl hostname overwrite 之后就不需要在环境里面搞这段了

~~您可能还需要在 hosts 文件中增加节点域名和 IP 的映射关系~~

**Comment by davidkhala** [comment]
@SamYuan1990 在做这个的途中发现logging的策略比较模糊，可能需要清理下，可以另外开issue么？

**Comment by SamYuan1990** [comment]
> @SamYuan1990 在做这个的途中发现logging的策略比较模糊，可能需要清理下，可以另外开issue么？

好的，欢迎

File: Dockerfile [hunk 3]
```diff
@@ -8,6 +8,7 @@ ENV GOPROXY=https://goproxy.cn,direct
 ENV export GOSUMDB=off
 
 COPY . .
+RUN go mod vendor
```
Comment (SamYuan1990) [inline Dockerfile:11]: ? why we need `go mod vendor` here?

File: pkg/infra/basic/config.go (hunk 2 above)
Comment (SamYuan1990) [reply pkg/infra/basic/config.go:65]: at my point of view, ssl hostname overwrite still need you to change `/etc/host` if needed....

File: test/integration-test.sh [hunk 4]
```diff
@@ -13,15 +13,12 @@ case $1 in
...
     cd ./fabric-samples/
     git checkout release-1.4
     cd ./first-network
-    # 1.4.10
-    echo y | ./byfn.sh up -i 1.4.10
-    # comments here for 1.4.8 work around as docker image issue.
-    # docker pull hyperledger/fabric-orderer:amd64-1.4  
-    # Error response from daemon: manifest for hyperledger/fabric-orderer:amd64-1.4 not found: manifest unknown: manifest unknown
+
+    echo y | ./byfn.sh up -i 1.4.12
```
Comment (SamYuan1990) [inline test/integration-test.sh:21]: I suggest change back to `1.4.10`
is 1.4 still LTS?
are we going to upgrade CI changes with this pr?

File: Dockerfile (hunk 3 above)
Comment (davidkhala) [reply Dockerfile:11]: 解决依赖冲突的时候加入的代码，刚测试过，由于go build的时候也会做类似的事情，所以多余
将此回滚

File: pkg/infra/basic/config.go (hunk 2 above)
Comment (davidkhala) [reply pkg/infra/basic/config.go:65]: /etc/host/ can be another layer if user want, but not mandatory for cheating grpc

File: test/integration-test.sh (hunk 4 above)
Comment (davidkhala) [reply test/integration-test.sh:21]: 1.4 is no more LTS, would you consider a removal for this?

Comment (SamYuan1990) [reply test/integration-test.sh:21]: So far not, as it is previous TLS, and I make a PR for supporting at https://github.com/Hyperledger-TWGC/tape/pull/250
but honestly speaking, for my point of view, as 1.4 is no previous LTS, we don't need to bump up version in CI in high frequency.

Comment (davidkhala) [reply test/integration-test.sh:21]: 1.4.12 should be last version of release-1.4, I think no further bump up in future.

**Comment by davidkhala** [comment]
@SamYuan1990 这个CI出的error让我想请教下
为何在文档当中我们都使用的是...msp/keystore/priv_sk
但是实际上存在的文件是：...msp/keystore/6e8f3e24524194db05ee8bd0b8989b6ba6551bad83f8cf4220d167221c82de36_sk

更为惊奇的是虽然 msp/keystore/priv_sk不存在，但是master上的1_4测试竟然也可以通过。

**Comment by davidkhala** [comment]
> @SamYuan1990 这个CI出的error让我想请教下 为何在文档当中我们都使用的是...msp/keystore/priv_sk 但是实际上存在的文件是：...msp/keystore/6e8f3e24524194db05ee8bd0b8989b6ba6551bad83f8cf4220d167221c82de36_sk
> 
> 更为惊奇的是虽然 msp/keystore/priv_sk不存在，但是master上的1_4测试竟然也可以通过。

了解到priv_sk是tape一直以来用的格式，也是一直以来cryptogen生成的格式

**Review by SamYuan1990 — APPROVED** [review]
LGTM

**Pull request merged**